README.md
# Global Mercator

[![Build Status](https://travis-ci.org/DenisCarriere/global-mercator.svg?branch=master)](https://travis-ci.org/DenisCarriere/global-mercator)
[![npm version](https://badge.fury.io/js/global-mercator.svg)](https://badge.fury.io/js/global-mercator)
[![Coverage Status](https://coveralls.io/repos/github/DenisCarriere/global-mercator/badge.svg?branch=master)](https://coveralls.io/github/DenisCarriere/global-mercator?branch=master)
[![MIT licensed](https://img.shields.io/badge/license-MIT-blue.svg)](https://raw.githubusercontent.com/DenisCarriere/global-mercator/master/LICENSE)
[![ES5](https://camo.githubusercontent.com/d341caa63123c99b79fda7f8efdc29b35f9f2e70/68747470733a2f2f696d672e736869656c64732e696f2f62616467652f65732d352d627269676874677265656e2e737667)](http://kangax.github.io/compat-table/es5/)

[![Standard - JavaScript Style Guide](https://cdn.rawgit.com/feross/standard/master/badge.svg)](https://github.com/feross/standard)

A set of tools geospatial tools to help with TMS, Google (XYZ) Tiles.

> This library is insipered by GDAL2Tiles, Google Summer of Code 2007 & 2008.

Another great simplistic tile library is [`tilebelt`](https://github.com/mapbox/tilebelt).

## Install

**npm**

```bash
$ npm install --save global-mercator
```

**web**

```html
<script src="https://wzrd.in/standalone/global-mercator@latest"></script>
```

## Quickstart

```javascript
var globalMercator = require('global-mercator')
var tile = [10, 15, 8] // [x, y, zoom]
globalMercator.tileToBBox(tile)
// [ -165.937, -82.853, -164.531, -82.676 ]
```

## Features

| Function                                                  | Description                                                          |
| --------------------------------------------------------- | :------------------------------------------------------------------- |
| [lngLatToMeters(LngLat)](#lnglattometers)                 | Converts LngLat coordinates to Meters coordinates.                   |
| [metersToLngLat(Meters)](#meterstolnglat)                 | Converts Meters coordinates to LngLat coordinates.                   |
| [metersToPixels(Meters, zoom)](#meterstopixels)           | Converts Meters coordinates to Pixels coordinates.                   |
| [lngLatToTile(LngLat, zoom)](#lnglattotile)               | Converts LngLat coordinates to TMS Tile.                             |
| [lngLatToGoogle(LngLat, zoom)](#lnglattogoogle)           | Converts LngLat coordinates to Google (XYZ) Tile.                    |
| [metersToTile(Meters, zoom)](#meterstotile)               | Converts Meters coordinates to TMS Tile.                             |
| [pixelsToMeters(Pixels)](#pixelstometers)                 | Converts Pixels coordinates to Meters coordinates.                   |
| [pixelsToTile(Pixels)](#pixelstotile)                     | Converts Pixels coordinates to TMS Tile.                             |
| [tileToBBoxMeters(tile)](#tiletobboxmeters)               | Converts TMS Tile to bbox in Meters coordinates.                     |
| [tileToBBox(tile)](#tiletobbox)                           | Converts TMS Tile to bbox in LngLat coordinates.                     |
| [googleToBBoxMeters(google)](#googletobboxmeters)         | Converts Google (XYZ) Tile to bbox in Meters coordinates.            |
| [googleToBBox(google)](#googletobbox)                     | Converts Google (XYZ) Tile to bbox in LngLat coordinates.            |
| [tileToGoogle(tile)](#tiletogoogle)                       | Converts TMS Tile to Google (XYZ) Tile.                              |
| [googleToTile(google)](#googletotile)                     | Converts Google (XYZ) Tile to TMS Tile.                              |
| [googleToQuadkey(google)](#googletoquadkey)               | Converts Google (XYZ) Tile to Quadkey.                               |
| [tileToQuadkey(tile)](#tiletoquadkey)                     | Converts TMS Tile to QuadKey.                                        |
| [quadkeyToTile(quadkey)](#quadkeytotile)                  | Converts Quadkey to TMS Tile.                                        |
| [quadkeyToGoogle(quadkey)](#quadkeytogoogle)              | Converts Quadkey to Google (XYZ) Tile.                               |
| [hash(tile)](#hash)                                       | Hash tile for unique id key                                          |
| [validateTile(tile)](#validatetile)                       | Validates TMS Tile                                                   |
| [validateZoom(zoom)](#validatezoom)                       | Validates Zoom level                                                 |
| [validateLngLat(LngLat)](#validatelnglat)                 | Validates LngLat coordinates                                         |
| [validatePixels(Pixels)](#validatepixels)                 | Validates Pixels coordinates                                         |
| [maxBBox(BBox\[\])](#maxbbox)                             | Maximum extent of BBox                                               |
| [validTile(tile)](#validtile)                             | Valid Tile                                                           |
| [longitude(degree)](#longitude)                           | Modifies a Longitude to fit within +/-180 degrees.                   |
| [latitude(degree)](#latitude)                             | Modifies a Latitude to fit within +/-90 degrees.                     |
| [pointToTile(lnglat, zoom)](#pointtotile)                 | Get the tile for a point at a specified zoom level                   |
| [pointToTileFraction(lnglat, zoom)](#pointtotilefraction) | Get the precise fractional tile location for a point at a zoom level |
| [wrapTile(tile)](#wraptile)                               | Handles tiles which crosses the 180th meridian                       |
| [bboxToTile(bbox)](#bboxtotile)                           | Get the smallest tile to cover a bbox                                |

## API

<!-- Generated by documentation.js. Update this documentation by updating the source code. -->

### hash

Hash tile for unique id key

**Parameters**

-   `tile` **Tile** [x, y, z]

**Examples**

```javascript
var id = globalMercator.hash([312, 480, 4])
//=5728
```

Returns **[number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number)** hash

### pointToTile

Get the tile for a point at a specified zoom level
<https://github.com/mapbox/tilebelt>

**Parameters**


CHANGELOG.md

# Changelog

## 3.0.3 - 2017-09-29

- Drop strict `[number, number]` for LngLat inputs

## 3.0.0 - 2017-09-29

- Added `bboxToTile`
- Support ES modules

## 2.8.4 - 2017-08-02

- Update docs to match https://wzrd.in
- Drop Tile Y wrapping

## 2.8.1 - 2017-07-19

- Add `wrapTile` method
- Include tests side by side with `tilebelt`
- Fixed issue with `tileToQuadkey`
- Drop rollup setup
- Add Git URL repo to `package.json`
- Added `longitude` & `latitude` as exposed methods
- Dropped `bbox-dateline` from dependencies
- Remove Jest in favor of a minimalistic test suite (Tap)

## 2.7.0 - 2017-04-19

- Add `validTile` method

## 2.6.0 - 2017-03-20

- Fix LatLng validation (issue with lat being a lng)

## 2.5.0 - 2017-03-18

- Add minified build using Uglifyjs
- Improve lat & lng validation
- Add automated Table of Contents

## 2.4.0 - 2017-03-10

- Remove `throw Error` in Lat & Lng validation process
- Handle Lat & Lng with +/-180 degrees

## 2.3.0 - 2017-03-06

- Add `validate` as optional `false` param

## 2.2.0 - 2017-03-04

- Bundle to UMD
- Modify rollup config
- Output browser bundle to `docs/`

## 2.1.0 - 2017-02-21

- Added `maxBBox` method
- Added Benchmark performance testing

## 2.0.0 - 2017-02.18

- Convert to ES5
- Convert project to support Rollup
- Add `tileSize` @param
- Dropped Grid methods (BREAKS backwards compatibility - Major release 2.X)
- Support browser, nodejs & commonjs

## 1.9.0 - 2017-02-17

- Update fixed bbox if zoom level = 0
- Add validation to latlng
- Add validation to tiles

## 1.8.0 - 2017-01-30

- Publish web browser support for ES6.

## 1.7.0 - 2017-01-24

- Refactoring entire codebase to standard Javascript

## 1.6.0 - 2017-01-05

- BBox reduced decimal precision to 6
- bboxToCenter toFixed(6)
- Lat & Lng decimal toFixed(6)
- Meters toFixed(1)

## 1.5.0 - 2016-12-25

- Add bboxToCenter & hash function
- Enforce strict [number, number] instead of number[]

## 1.4.0 - 2016-12-4

- Replace test tools for Jest
- Added Circle CI testing

## 1.3.5 - 2016-10-25

Reduce library size by 430% (350KB to 80KB)

## 1.3.1 - 2016-10-14

- Update typings to allow `number[]`
- Enforcing validation with `validateLngLat`, `validateMeters` & `validateTile`
- Reviewed entire documentation
- Dropped zoom `@param` from functions with only LatLng & Meters.

```javascript
lngLatToMeters([lng, lat, zoom])
// Changed to
lngLatToMeters([lng, lat])

metersToLngLat([x, y, zoom])
// Changed to
metersToLngLat([x, y])
```

## 1.2.0 - 2016-10-13

Added new features:

- `grid` - Creates an Iterator of Tiles from a given BBox
- `gridBulk` - Creates a bulk Iterator of Tiles from a given BBox
- `gridLevels` - Creates a grid level pattern of arrays
- `gridCount` - Counts the total amount of tiles from a given BBox

## 1.1.0 - 2016-10-03

- Remove Default export
- Modules compiles only to ES6

## 1.0.0 - 2016-10-01

First Stable release was created

index.js
var originShift = 2 * Math.PI * 6378137 / 2.0
var d2r = Math.PI / 180

export function initialResolution (tileSize) {
  tileSize = tileSize || 256
  return 2 * Math.PI * 6378137 / tileSize
}

/**
 * Hash tile for unique id key
 *
 * @param {Tile} tile [x, y, z]
 * @returns {number} hash
 * @example
 * var id = globalMercator.hash([312, 480, 4])
 * //=5728
 */
export function hash (tile) {
  var x = tile[0]
  var y = tile[1]
  var z = tile[2]
  return (1 << z) * ((1 << z) + x) + y
}

/**
 * Get the tile for a point at a specified zoom level
 * https://github.com/mapbox/tilebelt
 *
 * @param {[number, number]} lnglat [Longitude, Latitude]
 * @param {number} zoom Zoom level
 * @param {boolean} [validate=true] validates LatLng coordinates
 * @returns {Google} Google (XYZ) Tile
 * @example
 * var tile = globalMercator.pointToTile([1, 1], 12)
 * //= [ 2059, 2036, 12 ]
 */
export function pointToTile (lnglat, zoom, validate) {
  var tile = pointToTileFraction(lnglat, zoom, validate)
  tile[0] = Math.floor(tile[0])
  tile[1] = Math.floor(tile[1])
  return tile
}

/**
 * Get the precise fractional tile location for a point at a zoom level
 * https://github.com/mapbox/tilebelt
 *
 * @name pointToTileFraction
 * @param {[number, number]} lnglat [Longitude, Latitude]
 * @param {number} zoom Zoom level
 * @param {boolean} [validate=true] validates LatLng coordinates
 * @returns {Google} Google (XYZ) Tile
 * @example
 * var tile = globalMercator.pointToTileFraction([1, 1], 12)
 * //= [ 2059.3777777777777, 2036.6216445333432, 12 ]
 */
export function pointToTileFraction (lnglat, zoom, validate) {
  // lnglat = validateLngLat(lnglat, validate)
  var z = zoom
  var lon = longitude(lnglat[0])
  var lat = latitude(lnglat[1])
  var sin = Math.sin(lat * d2r)
  var z2 = Math.pow(2, z)
  var x = z2 * (lon / 360 + 0.5)
  var y = z2 * (0.5 - 0.25 * Math.log((1 + sin) / (1 - sin)) / Math.PI)
  return validateTile([x, y, z], validate)
}

/**
 * Converts BBox to Center
 *
 * @param {BBox} bbox - [west, south, east, north] coordinates
 * @return {LngLat} center
 * @example
 * var center = globalMercator.bboxToCenter([90, -45, 85, -50])
 * //= [ 87.5, -47.5 ]
 */
export function bboxToCenter (bbox) {
  var west = bbox[0]
  var south = bbox[1]
  var east = bbox[2]
  var north = bbox[3]
  var lng = (west - east) / 2 + east
  var lat = (south - north) / 2 + north
  lng = Number(lng.toFixed(6))
  lat = Number(lat.toFixed(6))
  return [lng, lat]
}

/**
 * Converts LngLat coordinates to Meters coordinates.
 *
 * @param {[number, number]} lnglat [Longitude, Latitude]
 * @param {boolean} [validate=true] validates LatLng coordinates
 * @returns {Meters} Meters coordinates
 * @example
 * var meters = globalMercator.lngLatToMeters([126, 37])
 * //=[ 14026255.8, 4439106.7 ]
 */
export function lngLatToMeters (lnglat, validate) {
  lnglat = validateLngLat(lnglat, validate)
  var lng = lnglat[0]
  var lat = lnglat[1]
  var x = lng * originShift / 180.0
  var y = Math.log(Math.tan((90 + lat) * Math.PI / 360.0)) / (Math.PI / 180.0)
  y = y * originShift / 180.0
  x = Number(x.toFixed(1))
  y = Number(y.toFixed(1))
  return [x, y]
}

/**
 * Converts Meters coordinates to LngLat coordinates.
 *
 * @param {Meters} meters Meters in Mercator [x, y]
 * @returns {LngLat} LngLat coordinates
 * @example
 * var lnglat = globalMercator.metersToLngLat([14026255, 4439106])
 * //=[ 126, 37 ]
 */
export function metersToLngLat (meters) {
  var x = meters[0]
  var y = meters[1]
  var lng = (x / originShift) * 180.0
  var lat = (y / originShift) * 180.0
  lat = 180 / Math.PI * (2 * Math.atan(Math.exp(lat * Math.PI / 180.0)) - Math.PI / 2.0)
  lng = Number(lng.toFixed(6))
  lat = Number(lat.toFixed(6))
  return [lng, lat]
}

/**
 * Converts Meters coordinates to Pixels coordinates.
 *
 * @param {Meters} meters Meters in Mercator [x, y]
 * @param {number} zoom Zoom level
 * @param {number} [tileSize=256] Tile size
 * @returns {Pixels} Pixels coordinates
 * @example
 * var pixels = globalMercator.metersToPixels([14026255, 4439106], 13)
 * //=[ 1782579.1, 1280877.3, 13 ]
 */
export function metersToPixels (meters, zoom, tileSize) {
  var x = meters[0]
  var y = meters[1]
  var res = resolution(zoom, tileSize)
  var px = (x + originShift) / res
  var py = (y + originShift) / res
  return [px, py, zoom]
}

/**
 * Converts LngLat coordinates to TMS Tile.
 *
 * @param {[number, number]} lnglat [Longitude, Latitude]
 * @param {number} zoom Zoom level
 * @param {boolean} [validate=true] validates LatLng coordinates
 * @returns {Tile} TMS Tile
 * @example
 * var tile = globalMercator.lngLatToTile([126, 37], 13)
 * //=[ 6963, 5003, 13 ]
 */
export function lngLatToTile (lnglat, zoom, validate) {
  lnglat = validateLngLat(lnglat, validate)
  var meters = lngLatToMeters(lnglat)
  var pixels = metersToPixels(meters, zoom)
  return pixelsToTile(pixels)
}

/**
 * Converts LngLat coordinates to Google (XYZ) Tile.
 *
 * @param {[number, number]} lnglat [Longitude, Latitude]
 * @param {number} zoom Zoom level
 * @param {boolean} [validate=true] validates LatLng coordinates
 * @returns {Google} Google (XYZ) Tile
 * @example
 * var google = globalMercator.lngLatToGoogle([126, 37], 13)
 * //=[ 6963, 3188, 13 ]
 */
export function lngLatToGoogle (lnglat, zoom, validate) {
  lnglat = validateLngLat(lnglat, validate)

  if (zoom === 0) {
    return [0, 0, 0]
  }
  var tile = lngLatToTile(lnglat, zoom)
  return tileToGoogle(tile)
}

/**
 * Converts Meters coordinates to TMS Tile.
 *
 * @param {Meters} meters Meters in Mercator [x, y]
 * @param {number} zoom Zoom level
 * @returns {Tile} TMS Tile
 * @example
 * var tile = globalMercator.metersToTile([14026255, 4439106], 13)
 * //=[ 6963, 5003, 13 ]
 */
export function metersToTile (meters, zoom) {
  if (zoom === 0) {
    return [0, 0, 0]
  }
  var pixels = metersToPixels(meters, zoom)
  return pixelsToTile(pixels)
}

/**
 * Converts Pixels coordinates to Meters coordinates.
 *
 * @param {Pixels} pixels Pixels [x, y, zoom]
 * @param {number} [tileSize=256] Tile size
 * @returns {Meters} Meters coordinates
 * @example
 * var meters = globalMercator.pixelsToMeters([1782579, 1280877, 13])
 * //=[ 14026252.0, 4439099.5 ]
 */
export function pixelsToMeters (pixels, tileSize) {
  var px = pixels[0]
  var py = pixels[1]
  var zoom = pixels[2]
  var res = resolution(zoom, tileSize)
  var mx = px * res - originShift
  var my = py * res - originShift
  mx = Number(mx.toFixed(1))
  my = Number(my.toFixed(1))
  return [mx, my]
}

/**
 * Converts Pixels coordinates to TMS Tile.
 *
 * @param {Pixels} pixels Pixels [x, y, zoom]
 * @param {number} [tileSize=256] Tile size
 * @param {boolean} [validate=true] validates Pixels coordinates
 * @returns {Tile} TMS Tile
 * @example
 * var tile = globalMercator.pixelsToTile([1782579, 1280877, 13])
 * //=[ 6963, 5003, 13 ]
 */
export function pixelsToTile (pixels, tileSize, validate) {
  tileSize = tileSize || 256
  var px = pixels[0]
  var py = pixels[1]
  var zoom = pixels[2]
  if (zoom === 0) return [0, 0, 0]

  validateZoom(zoom, validate)
  var tx = Math.ceil(px / tileSize) - 1
  var ty = Math.ceil(py / tileSize) - 1
  if (tx < 0) tx = 0
  if (ty < 0) ty = 0
  return [tx, ty, zoom]
}

/**
 * Converts TMS Tile to bbox in Meters coordinates.
 *
 * @param {Tile} tile Tile [x, y, zoom]
 * @param {number} x TMS Tile X
 * @param {number} y TMS Tile Y
 * @param {number} zoom Zoom level
 * @param {number} [tileSize=256] Tile size
 * @param {boolean} [validate=true] validates Tile
 * @returns {BBox} bbox extent in [minX, minY, maxX, maxY] order
 * @example
 * var bbox = globalMercator.tileToBBoxMeters([6963, 5003, 13])
 * //=[ 14025277.4, 4437016.6, 14030169.4, 4441908.5 ]
 */
export function tileToBBoxMeters (tile, tileSize, validate) {
  validateTile(tile, validate)

  tileSize = tileSize || 256
  var tx = tile[0]
  var ty = tile[1]
  var zoom = tile[2]
  var min = pixelsToMeters([tx * tileSize, ty * tileSize, zoom])
  var max = pixelsToMeters([(tx + 1) * tileSize, (ty + 1) * tileSize, zoom])
  return [min[0], min[1], max[0], max[1]]
}

/**
 * Converts TMS Tile to bbox in LngLat coordinates.
 *
 * @param {Tile} tile Tile [x, y, zoom]
 * @param {number} x TMS Tile X
 * @param {number} y TMS Tile Y
 * @param {number} zoom Zoom level
 * @param {boolean} [validate=true] validates Tile
 * @returns {BBox} bbox extent in [minX, minY, maxX, maxY] order
 * @example
 * var bbox = globalMercator.tileToBBox([6963, 5003, 13])
 * //=[ 125.991, 36.985, 126.035, 37.020 ]
 */
export function tileToBBox (tile, validate) {
  validateTile(tile, validate)

  var tx = tile[0]
  var ty = tile[1]
  var zoom = tile[2]
  if (zoom === 0) {
    return [-180, -85.051129, 180, 85.051129]
  }
  var bbox = tileToBBoxMeters([tx, ty, zoom])
  var mx1 = bbox[0]
  var my1 = bbox[1]
  var mx2 = bbox[2]
  var my2 = bbox[3]
  var min = metersToLngLat([mx1, my1, zoom])
  var max = metersToLngLat([mx2, my2, zoom])
  return [min[0], min[1], max[0], max[1]]
}

/**
 * Converts Google (XYZ) Tile to bbox in Meters coordinates.
 *
 * @param {Google} google Google [x, y, zoom]
 * @returns {BBox} bbox extent in [minX, minY, maxX, maxY] order
 * @example
 * var bbox = globalMercator.googleToBBoxMeters([6963, 3188, 13])
 * //=[ 14025277.4, 4437016.6, 14030169.4, 4441908.5 ]
 */
export function googleToBBoxMeters (google) {
  var Tile = googleToTile(google)
  return tileToBBoxMeters(Tile)
}

/**
 * Converts Google (XYZ) Tile to bbox in LngLat coordinates.
 *
 * @param {Google} google Google [x, y, zoom]
 * @returns {BBox} bbox extent in [minX, minY, maxX, maxY] order
 * @example
 * var bbox = globalMercator.googleToBBox([6963, 3188, 13])
 * //=[ 125.991, 36.985, 126.035, 37.020 ]
 */
export function googleToBBox (google) {
  var Tile = googleToTile(google)
  return tileToBBox(Tile)
}

/**
 * Converts TMS Tile to Google (XYZ) Tile.
 *
 * @param {Tile} tile Tile [x, y, zoom]
 * @param {boolean} [validate=true] validates Tile
 * @returns {Google} Google (XYZ) Tile
 * @example
 * var google = globalMercator.tileToGoogle([6963, 5003, 13])
 * //=[ 6963, 3188, 13 ]
 */
export function tileToGoogle (tile, validate) {
  validateTile(tile, validate)

  var tx = tile[0]
  var ty = tile[1]
  var zoom = tile[2]
  if (zoom === 0) {
    return [0, 0, 0]
  }
  var x = tx
  var y = (Math.pow(2, zoom) - 1) - ty
  return [x, y, zoom]
}

/**
 * Converts Google (XYZ) Tile to TMS Tile.
 *
 * @param {Google} google Google [x, y, zoom]
 * @returns {Tile} TMS Tile
 * @example
 * var tile = globalMercator.googleToTile([6963, 3188, 13])
 * //=[ 6963, 5003, 13 ]
 */
export function googleToTile (google) {
  var x = google[0]
  var y = google[1]
  var zoom = google[2]
  var tx = x
  var ty = Math.pow(2, zoom) - y - 1
  return [tx, ty, zoom]
}

/**
 * Converts Google (XYZ) Tile to Quadkey.
 *
 * @param {Google} google Google [x, y, zoom]
 * @returns {string} Microsoft's Quadkey schema
 * @example
 * var quadkey = globalMercator.googleToQuadkey([6963, 3188, 13])
 * //='1321102330211'
 */
export function googleToQuadkey (google) {
  var Tile = googleToTile(google)
  return tileToQuadkey(Tile)
}

/**
 * Converts TMS Tile to QuadKey.
 *
 * @param {Tile} tile Tile [x, y, zoom]
 * @param {boolean} [validate=true] validates Tile
 * @returns {string} Microsoft's Quadkey schema
 * @example
 * var quadkey = globalMercator.tileToQuadkey([6963, 5003, 13])
 * //='1321102330211'
 */
export function tileToQuadkey (tile, validate) {
  validateTile(tile, validate)

  var tx = tile[0]
  var ty = tile[1]
  var zoom = tile[2]
  // Zoom 0 does not exist for Quadkey
  if (zoom === 0) {
    return ''
  }
  var quadkey = ''
  ty = (Math.pow(2, zoom) - 1) - ty
  range(zoom, 0, -1).map(function (i) {
    var digit = 0
    var mask = 1 << (i - 1)
    if ((tx & mask) !== 0) {
      digit += 1
    }
    if ((ty & mask) !== 0) {
      digit += 2
    }
    quadkey = quadkey.concat(digit)
  })
  return quadkey
}

/**
 * Converts Quadkey to TMS Tile.
 *
 * @param {string} quadkey Microsoft's Quadkey schema
 * @returns {Tile} TMS Tile
 * @example
 * var tile = globalMercator.quadkeyToTile('1321102330211')
 * //=[ 6963, 5003, 13 ]
 */
export function quadkeyToTile (quadkey) {
  var Google = quadkeyToGoogle(quadkey)
  return googleToTile(Google)
}

/**
 * Converts Quadkey to Google (XYZ) Tile.
 *
 * @param {string} quadkey Microsoft's Quadkey schema
 * @returns {Google} Google (XYZ) Tile
 * @example
 * var google = globalMercator.quadkeyToGoogle('1321102330211')
 * //=[ 6963, 3188, 13 ]
 */
export function quadkeyToGoogle (quadkey) {
  var x = 0
  var y = 0
  var zoom = quadkey.length
  range(zoom, 0, -1).map(function (i) {
    var mask = 1 << (i - 1)
    switch (parseInt(quadkey[zoom - i], 0)) {
      case 0:
        break
      case 1:
        x += mask
        break
      case 2:
        y += mask
        break
      case 3:
        x += mask
        y += mask
        break
      default:
        throw new Error('Invalid Quadkey digit sequence')
    }
  })
  return [x, y, zoom]
}

/**
 * Converts BBox from LngLat coordinates to Meters coordinates
 *
 * @param {BBox} bbox extent in [minX, minY, maxX, maxY] order
 * @returns {BBox} bbox extent in [minX, minY, maxX, maxY] order
 * @example
 * var meters = globalMercator.bboxToMeters([ 125, 35, 127, 37 ])
 * //=[ 13914936.3, 4163881.1, 14137575.3, 4439106.7 ]
 */
export function bboxToMeters (bbox) {
  var min = lngLatToMeters([bbox[0], bbox[1]])
  var max = lngLatToMeters([bbox[2], bbox[3]])
  return [min[0], min[1], max[0], max[1]]
}

/**
 * Validates TMS Tile.
 *
 * @param {Tile} tile Tile [x, y, zoom]
 * @param {boolean} [validate=true] validates Tile
 * @throws {Error} Will throw an error if TMS Tile is not valid.
 * @returns {Tile} TMS Tile
 * @example
 * globalMercator.validateTile([60, 80, 12])
 * //=[60, 80, 12]
 * globalMercator.validateTile([60, -43, 5])
 * //= Error: Tile <y> must not be less than 0
 * globalMercator.validateTile([25, 60, 3])
 * //= Error: Illegal parameters for tile
 */
export function validateTile (tile, validate) {
  var tx = tile[0]
  var ty = tile[1]
  var zoom = tile[2]
  if (validate === false) return tile
  if (zoom === undefined || zoom === null) throw new Error('<zoom> is required')
  if (tx === undefined || tx === null) throw new Error('<x> is required')
  if (ty === undefined || ty === null) throw new Error('<y> is required')

  // Adjust values of tiles to fit within tile scheme
  zoom = validateZoom(zoom)
  tile = wrapTile(tile)

  // // Check to see if tile is valid based on the zoom level
  // // Currently impossible to hit since WrapTile handles this error
  // // will keep this test commented out in case it doesnt handle it
  // var maxCount = Math.pow(2, zoom)
  // if (tile[0] >= maxCount || tile[1] >= maxCount) throw new Error('Illegal parameters for tile')
  return tile
}

/**
 * Wrap Tile -- Handles tiles which crosses the 180th meridian or 90th parallel
 *
 * @param {[number, number, number]} tile Tile
 * @param {number} zoom Zoom Level
 * @returns {[number, number, number]} Wrapped Tile
 * @example
 * globalMercator.wrapTile([0, 3, 2])
 * //= [0, 3, 2] -- Valid Tile X
 * globalMercator.wrapTile([4, 2, 2])
 * //= [0, 2, 2] -- Tile 4 does not exist, wrap around to TileX=0
 */
export function wrapTile (tile) {
  var tx = tile[0]
  var ty = tile[1]
  var zoom = tile[2]

  // Maximum tile allowed
  // zoom 0 => 1
  // zoom 1 => 2
  // zoom 2 => 4
  // zoom 3 => 8
  var maxTile = Math.pow(2, zoom)

  // Handle Tile X
  tx = tx % maxTile
  if (tx < 0) tx = tx + maxTile

  return [tx, ty, zoom]
}

/**
 * Validates Zoom level
 *
 * @param {number} zoom Zoom level
 * @param {boolean} [validate=true] validates Zoom level
 * @throws {Error} Will throw an error if zoom is not valid.
 * @returns {number} zoom Zoom level
 * @example
 * globalMercator.validateZoom(12)
 * //=12
 * globalMercator.validateZoom(-4)
 * //= Error: <zoom> cannot be less than 0
 * globalMercator.validateZoom(32)
 * //= Error: <zoom> cannot be greater than 30
 */
export function validateZoom (zoom) {
  if (zoom === false) return zoom
  if (zoom === undefined || zoom === null) { throw new Error('<zoom> is required') }
  if (zoom < 0) { throw new Error('<zoom> cannot be less than 0') }
  if (zoom > 32) { throw new Error('<zoom> cannot be greater than 32') }
  return zoom
}

/**
 * Validates LngLat coordinates
 *
 * @param {[number, number]} lnglat [Longitude, Latitude]
 * @param {boolean} [validate=true] validates LatLng coordinates
 * @throws {Error} Will throw an error if LngLat is not valid.
 * @returns {LngLat} LngLat coordinates
 * @example
 * globalMercator.validateLngLat([-115, 44])
 * //= [ -115, 44 ]
 * globalMercator.validateLngLat([-225, 44])
 * //= Error: LngLat [lng] must be within -180 to 180 degrees
 */
export function validateLngLat (lnglat, validate) {
  if (validate === false) return lnglat

  var lng = longitude(lnglat[0])
  var lat = latitude(lnglat[1])

  // Global Mercator does not support latitudes within 85 to 90 degrees
  if (lat > 85) lat = 85
  if (lat < -85) lat = -85
  return [lng, lat]
}

/**
 * Retrieve resolution based on zoom level
 *
 * @private
 * @param {number} zoom zoom level
 * @param {number} [tileSize=256] Tile size
 * @returns {number} resolution
 * @example
 * var res = globalMercator.resolution(13)
 * //=19.109257071294063
 */
export function resolution (zoom, tileSize) {
  return initialResolution(tileSize) / Math.pow(2, zoom)
}

/**
 * Generate an integer Array containing an arithmetic progression.
 *
 * @private
 * @param {number} [start=0] Start
 * @param {number} stop Stop
 * @param {number} [step=1] Step
 * @returns {number[]} range
 * @example
 * globalMercator.range(3)
 * //=[ 0, 1, 2 ]
 * globalMercator.range(3, 6)
 * //=[ 3, 4, 5 ]
 * globalMercator.range(6, 3, -1)
 * //=[ 6, 5, 4 ]
 */
export function range (start, stop, step) {
  if (stop == null) {
    stop = start || 0
    start = 0
  }
  if (!step) {
    step = stop < start ? -1 : 1
  }
  var length = Math.max(Math.ceil((stop - start) / step), 0)
  var range = Array(length)
  for (var idx = 0; idx < length; idx++, start += step) {
    range[idx] = start
  }
  return range
}

/**
 * Maximum extent of BBox
 *
 * @param {BBox|BBox[]} array BBox [west, south, east, north]
 * @returns {BBox} Maximum BBox
 * @example
 * var bbox = globalMercator.maxBBox([[-20, -30, 20, 30], [-110, -30, 120, 80]])
 * //=[-110, -30, 120, 80]
 */
export function maxBBox (array) {
  if (!array) throw new Error('array is required')

  // Single BBox
  if (array && array[0] && array.length === 4 && array[0][0] === undefined) {
    return array
  }

  // Multiple BBox
  if (array && array[0] && array[0][0] !== undefined) {
    var west = array[0][0]
    var south = array[0][1]
    var east = array[0][2]
    var north = array[0][3]

    array.map(function (bbox) {
      if (bbox[0] < west) { west = bbox[0] }
      if (bbox[1] < south) { south = bbox[1] }
      if (bbox[2] > east) { east = bbox[2] }
      if (bbox[3] > north) { north = bbox[3] }
    })
    return [west, south, east, north]
  }
}

/**
 * Valid TMS Tile
 *
 * @param {Tile} tile Tile [x, y, zoom]
 * @returns {boolean} valid tile true/false
 * @example
 * globalMercator.validTile([60, 80, 12])
 * //= true
 * globalMercator.validTile([60, -43, 5])
 * //= false
 * globalMercator.validTile([25, 60, 3])
 * //= false
 */
export function validTile (tile) {
  try {
    validateTile(tile)
    return true
  } catch (e) {
    return false
  }
}

/**
 * Modifies a Latitude to fit within +/-90 degrees.
 *
 * @param {number} lat latitude to modify
 * @returns {number} modified latitude
 * @example
 * globalMercator.latitude(100)
 * //= -80
 */
export function latitude (lat) {
  if (lat === undefined || lat === null) throw new Error('lat is required')

  // Latitudes cannot extends beyond +/-90 degrees
  if (lat > 90 || lat < -90) {
    lat = lat % 180
    if (lat > 90) lat = -180 + lat
    if (lat < -90) lat = 180 + lat
    if (lat === 0) lat = 0
  }
  return lat
}

/**
 * Modifies a Longitude to fit within +/-180 degrees.
 *
 * @param {number} lng longitude to modify
 * @returns {number} modified longitude
 * @example
 * globalMercator.longitude(190)
 * //= -170
 */
export function longitude (lng) {
  if (lng === undefined || lng === null) throw new Error('lng is required')

  // lngitudes cannot extends beyond +/-90 degrees
  if (lng > 180 || lng < -180) {
    lng = lng % 360
    if (lng > 180) lng = -360 + lng
    if (lng < -180) lng = 360 + lng
    if (lng === 0) lng = 0
  }
  return lng
}

/**
 * Get the smallest tile to cover a bbox
 *
 * @param {Array<number>} bbox BBox
 * @returns {Array<number>} tile Tile
 * @example
 * var tile = bboxToTile([-178, 84, -177, 85])
 * //=tile
 */
export function bboxToTile (bboxCoords) {
  var min = pointToTile([bboxCoords[0], bboxCoords[1]], 32)
  var max = pointToTile([bboxCoords[2], bboxCoords[3]], 32)
  var bbox = [min[0], min[1], max[0], max[1]]

  var z = getBboxZoom(bbox)
  if (z === 0) return [0, 0, 0]
  var x = bbox[0] >>> (32 - z)
  var y = bbox[1] >>> (32 - z)
  return [x, y, z]
}

function getBboxZoom (bbox) {
  var MAX_ZOOM = 28
  for (var z = 0; z < MAX_ZOOM; z++) {
    var mask = 1 << (32 - (z + 1))
    if (((bbox[0] & mask) !== (bbox[2] & mask)) ||
      ((bbox[1] & mask) !== (bbox[3] & mask))) {
      return z
    }
  }

  return MAX_ZOOM
}
